File: modules/java/lib/version/src/main/java/com/mooltiverse/oss/nyx/version/CompositeObjectIdentifier.java
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooltiverse.oss.nyx.version;

import java.util.List;

/**
 * A composite identifier that only accepts {@link SimpleIdentifier}s as children.
 */
class CompositeObjectIdentifier extends CompositeIdentifier {
    /**
     * Serial version UID to comply with {@link java.io.Serializable}
     */
    private static final long serialVersionUID = 1L;
    
    /**
     * Builds the identifier with the given nested identifiers and separator.
     *
     * @param separator the separator used for the string representation to separate children values
     * @param children the children of this composite identifier. It can't be {@code null} or contain {@code null} values
     *
     * @throws NullPointerException if the given list of children is {@code null} or contains {@code null} values
     * @throws IllegalArgumentException if the given list of children contains illegal values
     */
    protected CompositeObjectIdentifier(char separator, SimpleIdentifier... children) {
        super(separator, children);
    }

    /**
     * Builds the identifier with the given nested identifiers and the default separator.
     *
     * @param children the children of this composite identifier. It can't be {@code null} or contain {@code null} values
     *
     * @throws NullPointerException if the given list of children is {@code null} or contains {@code null} values
     * @throws IllegalArgumentException if the given list of children contains illegal values
     */
    protected CompositeObjectIdentifier(SimpleIdentifier... children) {
        super(children);
    }

    /**
     * Builds the identifier with the given nested identifiers and separator.
     *
     * @param separator the separator used for the string representation to separate children values
     * @param children the children of this composite identifier. It can't be {@code null} or contain {@code null} values
     *
     * @throws NullPointerException if the given list of children is {@code null} or contains {@code null} values
     * @throws IllegalArgumentException if the given list of children contains illegal values
     */
    protected CompositeObjectIdentifier(char separator, List<SimpleIdentifier> children) {
        super(separator, children);
    }

    /**
     * Builds the identifier with the given nested identifiers and the default separator.
     *
     * @param children the children of this composite identifier. It can't be {@code null} or contain {@code null} values
     *
     * @throws NullPointerException if the given list of children is {@code null} or contains {@code null} values
     * @throws IllegalArgumentException if the given list of children contains illegal values
     */
    protected CompositeObjectIdentifier(List<SimpleIdentifier> children) {
        super(children);
    }

    /**
     * Returns an identifier representing the specified String value. The string is split using the
     * default separator.
     *
     * @param multipleIdentifiers when {@code true} the given string is parsed as it (may) contain multiple
     * identifiers, separated by the default separator, so this method may yield to multiple identifiers.
     * When {@code false} the given string is expected to have a single identifier so if the given
     * string has multiple identifiers an exception is thrown.
     * @param s the string to parse
     * @param integersPolicy the flag specifying how to parse integers
     *
     * @return the new identifier representing the given string.
     *
     * @throws NullPointerException if the given string is {@code null}
     * @throws IllegalArgumentException if the given string contains illegal identifiers
     */
    static CompositeObjectIdentifier valueOf(boolean multipleIdentifiers, String s, UseIntegerIdentifiers integersPolicy) {
        return valueOf(multipleIdentifiers, s, DEFAULT_SEPARATOR, integersPolicy);
    }

     /**
     * Returns an identifier representing the specified String value.
     *
     * @param multipleIdentifiers when {@code true} the given string is parsed as it (may) contain multiple
     * identifiers, separated by the given separator, so this method may yield to multiple identifiers.
     * When {@code false} the given string is expected to have a single identifier so if the given
     * string has multiple identifiers an exception is thrown.
     * @param s the string to parse
     * @param separator the separator used for the string representation to separate children values
     * @param integersPolicy the flag specifying how to parse integers
     *
     * @return the new identifier representing the given string.
     *
     * @throws NullPointerException if the given string is {@code null}
     * @throws IllegalArgumentException if the given string contains illegal identifiers
     */
    static CompositeObjectIdentifier valueOf(boolean multipleIdentifiers, String s, char separator, UseIntegerIdentifiers integersPolicy) {
        if (multipleIdentifiers)
            return new CompositeObjectIdentifier(Parser.toIdentifiers(s, separator, integersPolicy));
        else return new CompositeObjectIdentifier(List.<SimpleIdentifier>of(Parser.toIdentifier(s, integersPolicy)));
    }

    /**
     * Returns the object at the given position (the first being at position 0)
     *
     * @param i the position to retrieve the value at (starting from 0)
     *
     * @return the object at the given position
     *
     * @throws IndexOutOfBoundsException if the given index is out of range
     */
    public Object get(int i) {
        return children.get(i).getValue();
    }
}

File: src/go/version/composite_object_identifier.go
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package version

import (
	"reflect" // https://pkg.go.dev/reflect
)

/*
A composite identifier that only accepts simpleIdentifiers as children.
*/
type compositeObjectIdentifier struct {
	// Extend compositeIdentifier by composition
	compositeIdentifier

	// The nested identifiers. This is redundant with the member from the nested compositeIdentifier
	// but it's typed so we can save methods with (up/down) casting.
	children []identifier
}

/*
Builds the identifier with the given nested identifiers and separator.
*/
func newCompositeObjectIdentifierWithSeparator(separator string, children ...identifier) (compositeObjectIdentifier, error) {
	identifier := compositeObjectIdentifier{}

	identifier.compositeIdentifier.separator = separator
	identifier.compositeIdentifier.children = children
	identifier.children = children

	return identifier, nil
}

/*
Builds the identifier with the given nested identifiers and the default separator.
*/
func newCompositeObjectIdentifier(children ...identifier) (compositeObjectIdentifier, error) {
	identifier := compositeObjectIdentifier{}

	identifier.compositeIdentifier.separator = DEFAULT_SEPARATOR
	identifier.compositeIdentifier.children = children
	identifier.children = children

	return identifier, nil
}

/*
Returns the object at the given position (the first being at position 0)
*/
func (coi compositeObjectIdentifier) get(i uint) interface{} {
	return coi.children[i].getValue()
}

/*
Tests the given object for equality against this instance.
*/
func (coi compositeObjectIdentifier) equals(obj interface{}) bool {
	if obj == nil {
		return false
	}
	otherIdentifier, ok := obj.(compositeObjectIdentifier)
	if !ok {
		return false
	}
	if coi.compositeIdentifier.separator != otherIdentifier.compositeIdentifier.separator {
		return false
	}
	return reflect.DeepEqual(coi.children, otherIdentifier.children)
}
